test/Spec/BeneficiariesFile.hs

```diff
@@ -1,35 +1,156 @@
-  let amt = normalize $ scientific (toInteger x) scale
-      s = "adfd87319bd09c9e3ea10b251ccb046f87c5440343157e348c3ac7bd " <> show amt <> " 1d6445ddeda578117f393848e685128f1e78ad0c4e48129c5964dc2e.testToken"
+oneAmount :: Assertion
+oneAmount = do
+  assertBool "Fail to parse if neither amounts are provided" . isLeft $ f False False
+  assertBool "Fail to parse if both amounts are provided" . isLeft $ f True True
+  where
+    f inFile inConfig =
+      parseContent (defaultConfig {dropAmount = defaultConfig.dropAmount <* guard inConfig, assetClass = Nothing})
+        . unwords
+        $ ["adfd87319bd09c9e3ea10b251ccb046f87c5440343157e348c3ac7bd"]
+          <> ["1000000" | inFile]
+          <> ["1d6445ddeda578117f393848e685128f1e78ad0c4e48129c5964dc2e.testToken"]
+
+oneAssetClass :: Assertion
+oneAssetClass = do
+  assertBool "Fail to parse if neither assetclasses are provided" . isLeft $ f False False
+  assertBool "Fail to parse if both assetclasses are provided" . isLeft $ f True True
+  where
+    f inFile inConfig =
+      parseContent (defaultConfig {dropAmount = Nothing, assetClass = defaultConfig.assetClass <* guard inConfig})
+        . unwords
+        $ ["adfd87319bd09c9e3ea10b251ccb046f87c5440343157e348c3ac7bd"]
+          <> ["1000000"]
+          <> ["1d6445ddeda578117f393848e685128f1e78ad0c4e48129c5964dc2e.testToken" | inFile]
+
+prop_TokenAmountTruncation :: TokenAmount -> Positive Int -> Property
+prop_TokenAmountTruncation (TokenAmount amt) (Positive dp) =
+  let s = "adfd87319bd09c9e3ea10b251ccb046f87c5440343157e348c3ac7bd " <> show amt <> " 1d6445ddeda578117f393848e685128f1e78ad0c4e48129c5964dc2e.testToken"
       shouldTruncate = dp + base10Exponent amt < 0
-      didTruncate = isLeft $ parseContent (defaultConfig {decimalPlaces = toInteger dp}) (pack s)
+      didTruncate = isLeft $ parseContent (defaultConfig {decimalPlaces = toInteger dp, dropAmount = Nothing, assetClass = Nothing}) (pack s)
    in didTruncate === shouldTruncate
 
+prop_Parse :: Beneficiaries -> Property
+prop_Parse (Beneficiaries (bens, config)) = either (error . unpack) (const $ property True) $ parseContent config bens
+
+-- TODO Generate [Beneficiary] directly, don't depend on parsing
+prop_Print :: Beneficiaries -> Property
+prop_Print (Beneficiaries (bens, config)) = either (error . unpack) (const $ property True) $ parseContent config bens >>= prettyContent config
+
+prop_Roundtrip1 :: Beneficiaries -> Property
+prop_Roundtrip1 (Beneficiaries (bens, config)) =
+  -- We can't check for equality immediately from the incoming Text, as information can be lost during parsing.
+  -- Thus we initially parse and then print, and then do the roundtrip, as parse >=> print should be idempotent
+  let printed = either (error . unpack) id $ parseContent config bens >>= prettyContent config
+      roundTrip = parseContent config >=> prettyContent config
+   in roundTrip printed === Right printed
+
+-- TODO Generate [Beneficiary] directly
```

Comment:
Will we be doing this in this PR?

Reply:
Nope, should I open some more issues WRT testing improvements?